include/nlohmann/detail/output/output_adapters.hpp

```diff
@@ -18,6 +18,8 @@ template<typename CharType> struct output_adapter_protocol
 {
     virtual void write_character(CharType c) = 0;
     virtual void write_characters(const CharType* s, std::size_t length) = 0;
+    virtual void write_characters_at(std::size_t position, const CharType* s, std::size_t length) = 0;
```

Comment:
I am a bit worried about this new function. So far, the output adapter interface was similar to an output iterator, but adding the `write_characters_at` function, it is more a random-access iterator. I understand that BSON needs to precede values with their byte size, but I fear that we may exclude output adapters in the future. If you have an idea how to realized writing BSON without `write_characters_at`, I would be happy.

If not we may keep it as is and cope with this in the future.

Reply:
I see your point. I guess there are several options available (beside keeping it as is):
1. Buffer internally (e.g. using and `std::vector<•>`), fix the size in the buffer at the end and then flush the buffer to the output iterator.
2. Find a way to estimate the size beforehand.

Option 1 is probably easier to implement but I dislike it because it requires a potentially very large buffer only to be able to fix a few bytes. Option 2 would be preferable, but I'm not sure yet how to implement it (given my limited familiarity with this code-base). This OTOH has the disadvantage of requiring an extra traversal of a potentially very large structure.

**Question:** Is there presently a convenient way to run a visitor (to calculate the necessary size) over the JSON object tree? 

I will think on the second option some more...

Reply:
Reworked the implementation to precompute the size of the BSON output, thus avoiding the additional random-access output-iterator requirement.

Reply:
Great!